exec/java-exec/src/main/java/org/apache/drill/exec/record/RawFragmentBatch.java

```diff
@@ -77,4 +83,46 @@ public long getByteCount() {
   public boolean isAckSent() {
     return ackSent.get();
   }
+
+  /**
+   * Transfer ownership of this DrillBuf to the target allocator. This is done for better memory
+   * accounting (that is, the operator should be charged with the body's Drillbuf memory).
+   *
+   * <p><b>NOTES -</b>
+   * <ul>
+   * <li>This operation is a NOOP when a) the current allocator (associated with the DrillBuf) is not the
+   *     owning allocator or b) the target allocator is already the owner
+   * <li>When transfer happens, a new RawFragmentBatch instance is allocated; this is done for proper
+   *     DrillBuf reference count accounting
+   * <li>The RPC handling code caches a reference to this RawFragmentBatch object instance; release()
+   *     calls should be routed to the previous DrillBuf
+   * </ul>
+   *
+   * @param targetAllocator target allocator
+   * @return a new {@link RawFragmentBatch} object instance on success (where the buffer ownership has
+   *         been switched to the target allocator); otherwise this operation is a NOOP (current instance
+   *         returned)
+   */
+  public RawFragmentBatch transferBodyOwnership(BufferAllocator targetAllocator) {
+    if (body == null) {
+      return this; // NOOP
+    }
+
+    if (!body.getLedger().isOwningLedger()
+     || body.getLedger().isOwner(targetAllocator)) {
+
+      return this;
+    }
+
+    int writerIndex               = body.writerIndex();
+    TransferResult transferResult = body.transferOwnership(targetAllocator);
```

Comment:
There is no check that the transfer is within new allocator limits, is it guaranteed to succeed?

Reply:
But what if it is an over limit?

Reply:
- After a call to getNext(), the operator code checks whether the transfer has succeeded
- So that an OOM condition is reported within the fragment
- When the change of ownership happens from parent to child, there should not be any OOM condition since this is internal account

For code clarity, I can improve the javadoc to ask the caller to check for an OOM condition (similarly to the BaseAlocator behavior).